Added func ref

tqwewe opened the pull request:
Added a function `ref(url string)` to change the current reference location of an existing Firebase reference.

tqwewe committed:
Added func ref

zabawaba99 commented:
Thanks for the PR, I can see where this would be quite useful and resolve #68. However, taking a look at the  [JavaScript](https://firebase.google.com/docs/reference/js/firebase.database.Database#ref), [Android](https://firebase.google.com/docs/reference/android/com/google/firebase/database/FirebaseDatabase.html#getReference(java.lang.String)), and [iOS](https://firebase.google.com/docs/reference/ios/firebasedatabase/api/reference/Classes/FIRDatabase#-referencewithpath) SDKs I wonder if the parameter of this method should be a `path` and not a `url`. It looks like they have both a `refWithpath` and a `refFromURL` in all of them. We could probably end up having both a `ref(path string) and a `refFromURL(url string)`. What do you think?

tqwewe commented:
After I pushed this edit, I realised the same thing but I've never contributed to a repo and it took me a while for me to figure out how to push this edit :P But yes I agree, it shouldn't be a URL... it should be the end of a URL ref("users/acidic9"). I'll try to make this edit and push it again!

tqwewe committed:
Added SetRef() and SetURL() functions

consulthys commented:
Nice! Is there any plan to merge this PR at some point?

tqwewe commented:
Actually on my own firebase.go file, I changed it to NewRef which actually copies the whole firebase reference , modifies the reference url and returns a new *Firebase. I did this in case of concurrent races that might occur if you keep changing the reference. Any thoughts on weather copying the reference would be better than just setting it?

This is the new code I have on my firebase.go file:
```go
func (fb *Firebase) NewRef(path string) *Firebase {
	newFB := fb.copy()

	parsedURL, err := _url.Parse(fb.url)
	if err != nil {
		panic(err)
	}
	if strings.HasPrefix(path, "/") {
		path = path[1:] // Remove prefix '/' if exists
	}
	newFB.url = parsedURL.Scheme + "://" + parsedURL.Host + "/" + path
	return newFB
}
```

zabawaba99 commented:
@Acidic9 I think copying the current current reference would be a good idea. It keeps the behavior consistent with the `child` function. As far as the naming though, I would prefer to keep it as `Ref`.

zabawaba99 commented:
Another note, we could probably consolidate that function to something like

```go
func (fb *Firebase) Ref(path string) *Firebase {
	newFB := fb.copy()
	url := strings.TrimPrefix(path, "/")
	newFB.url = sanitizeURL(url)
	return newFB
}
```

The `sanitizeURL` function handles the stripping of trailing slashes and also adds the `https://` scheme if it's missing.

tqwewe committed:
Changed SetRef to Ref Func

Removed the SetRef func replacing it with a new Ref() function to change an existing firebase reference.

tqwewe commented:
I just added this function, but actually, your version changes the whole url to the new reference.
For example, if I used the following code **with your function**:
```go
var result interface{}
usersRef := fb.Ref("users")
err := usersRef.Value(&result)
if err != nil {
	log.Fatal(err)
}
```
**It outputs the error:** 
`Get https://users/.json?auth=xxx: No such host is known.`

That's why in my function I used `_url.Parse(originalFirebaseURL)` then concatenated the url with the `scheme + "://" + host + "/" + newPath`.

I've rewritted the function to use the `strings.Trim` function.

```go
func (fb *Firebase) Ref(path string) *Firebase {
	newFB := fb.copy()
	parsedURL, err := _url.Parse(fb.url)
	if err != nil {
		panic(err)
	}
	newFB.url = parsedURL.Scheme + "://" + parsedURL.Host + "/" + strings.Trim(path, "/")
	return newFB
}
```

Suggestions still appreciated.

tqwewe committed:
Updated Ref() Func

The previous commit resulted in an incorrect url for the new Firebase reference.

tqwewe committed:
Updated Ref() Func

Changed strings.TrimPrefix to strings.Trim.

zabawaba99 commented:
@Acidic9 that's a good point. I completely overlooked that. 👍

zabawaba99 commented on firebase.go:
```diff
@@ -105,6 +105,22 @@ func (fb *Firebase) Unauth() {
+	newFB := fb.copy()
+	parsedURL, err := _url.Parse(fb.url)
+	if err != nil {
+		panic(err)
```
My only concern here is this panic. Could we return an error if we failed to parse the url instead? I want to avoid an unexpected panic happening if the user does something like


```go
fb := New('https://hello.com/world')
fb.SetURL(someVarThatIsNotAValidURL)
newRef := fb.Ref("hi")
```
After that, I think this is good to go.

tqwewe replied:
Okay, just fixed that, I only chose panic because it's so rare that the url will be unable to be parsed and it keeps it simple by only returning a single argument. But I agree that an error being returned could be better than panicing the whole program.

tqwewe committed:
Removed Panic from Ref() Func

Instead of returning only a *Firebase, the Ref function now also returns an error if the url is not able to be parsed correctly.

zabawaba99 commented:
Thanks for the contribution @Acidic9 !

consulthys commented:
Thanks, nice job guys !!

consulthys commented:
It would be nice to update the README file as well to show how to use this change.
